### PR TITLE
[18.0][MIG] mail_tracking: Migration to 18.0

### PR DESCRIPTION
**Pull request by trisdoan**
### Note
- missing commits from 14.0 are ported in https://github.com/trisdoan/social/pull/20

#### Backend
- `access_rights_uid` was removed from `_search()` changed in https://github.com/odoo/odoo/pull/171374/commits/21d023e6bc00793ac3b016cfdcafd05c9a81d95d
- `_prepare_outgoing_list` got extra param `mail_server` in https://github.com/odoo/odoo/pull/96008/commits/97ce0844ca0c3d531a7a2e2e47a867d89835b24e
- `_message_format` was converted to use store and `message_format_extras` was removed in https://github.com/odoo/odoo/commit/52df49eabb6e44ea6e0ed30b61e5a555af2e606a. As consequences, `_extras_to_store` is introduced
- `tools.ustr` is considered to be inefficient and flagged as deprecated in https://github.com/odoo/odoo/pull/173811/commits/64a5724e88de6532cf2a04e9243d16b079319ffd. Hence, is replaced by `exception_to_unicode` for the same effect.

#### Frontend
- `mail.messaging` service was removed in https://github.com/odoo/odoo/commit/683258b28107f44fba363e3b4e69f181cac3f80e
- `initMessagingCallback` is removed in https://github.com/odoo/odoo/commit/4f8087f03ad904b99f8f3580557a3ea744a97b76. So new entry point is created for store `onStarted` instead.
- mail.thread service was removed in  https://github.com/odoo/odoo/commit/65e5a7943de094fe538f2b978cb58c028a5694a1

### This changes
- Store helper was introduced in https://github.com/odoo/odoo/pull/150625/commits/c032ccfbaef91da0ee244f0e44e0e4e9d8d8d3fb. As consequences, failed data is now passed through Store.
   - As a side effect, it is also easier to do the search filter_failed(thread) with Store service
  
- As Odoo uses less Jquery, the module runs fine with vanilla. This impacts `mail_tracking/static/src/components/message_tracking/message_tracking.esm.js`

[[fixup! [MIG] mail_tracking: Migration to 18.0](https://github.com/trisdoan/social/pull/22/commits/832f9a64ae18103971ef51a775bd0ffa668c36c1)](https://github.com/trisdoan/social/pull/22/commits/85955995510bdd53c30e25fd9c395655f019bc01) (will squash later)
- Context: This [commit](https://github.com/OCA/social/pull/966/commits/b4ab1600fe0b464937ff3dc06ff7cbf3a29afe3a) porting from 12.0 implements permission constraint
  Approach is summarized as below:
    1. When do the searching, 
       - First, use the query to get result
       - Browse those id with the query + Remove any forbidden records
       - Return the accessible records in form of query, to continue the process
       
       => It could result in more queries -> performance cost
    
    2. When user read the records
      - The low-level method `_check_access` will discard forbidden records + raise exception

- Took this chance to
   - add security check test
   - replace % format with f-string

[[fixup! fixup! [MIG] mail_tracking: Migration to 18.0](https://github.com/trisdoan/social/pull/22/commits/498182b73c20acd9af2b2728b1b0f1cf561cc054)](https://github.com/trisdoan/social/pull/22/commits/b3df68e8bf6711d6c5355ada6687e076e34da6d7)
- Context: the FIXME was added in https://github.com/OCA/social/pull/624/commits/75604439061755da9df64f658a9951a85e196185, which was already solved in https://github.com/odoo/odoo/pull/49999/commits/7fe02c93e410f8e9d11c89f77c651ca4b93810ff

### Result
discuss_sidebar

failed_search_panel

failed_message_review

### CODE REVIEW

File: mail_tracking/static/src/components/message_tracking/message_tracking.esm.js [hunk 1]
```diff
@@ -9,7 +9,7 @@ export class MessageTracking extends Component {
         this.partner_trackings = useState(this.props.partner_trackings);
     }
     _onTrackingStatusClick(event) {
-        var tracking_email_id = $(event.currentTarget).data("tracking");
+        var tracking_email_id = event.currentTarget.dataset.tracking;
```
Comment (chaule97) [inline mail_tracking/static/src/components/message_tracking/message_tracking.esm.js]:
```suggestion
        const tracking_email_id = event.currentTarget.dataset.tracking;
```

File: mail_tracking/static/src/services/store_service_patch.esm.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    },
+
+    async filter_failed(thread) {
+        const {data} = await rpc(thread.getFetchRoute(), {
+            ...thread.getFetchParams(),
+        });
+
+        const messages = data["mail.message"].filter((message) => {
+            return message.is_failed_message;
+        });
```
Comment (chaule97) [inline mail_tracking/static/src/services/store_service_patch.esm.js]:
```suggestion
        const messages = data["mail.message"].filter(message => message.is_failed_message);
```

File: mail_tracking/static/src/components/failed_message/failed_message.esm.js [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    get thread() {
+        return this.props.message.thread;
+    }
+    get failed_recipients() {
+        const error_states = ["error", "rejected", "spam", "bounced", "soft-bounced"];
+        return this.message.partner_trackings.filter((message) => {
+            return error_states.includes(message.status);
+        });
```
Comment (chaule97) [inline mail_tracking/static/src/components/failed_message/failed_message.esm.js:97]:
```suggestion
        return this.message.partner_trackings.filter(message => error_states.includes(message.status));
```

File: mail_tracking/static/src/components/failed_message_review/failed_message_review.esm.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        });
+    }
+    get thread() {
+        return this.props.message.thread;
+    }
+    get failed_recipients() {
+        const error_states = ["error", "rejected", "spam", "bounced", "soft-bounced"];
+        return this.message.partner_trackings.filter((message) => {
+            return error_states.includes(message.status);
+        });
```
Comment (chaule97) [inline mail_tracking/static/src/components/failed_message_review/failed_message_review.esm.js:43]:
```suggestion
        return this.message.partner_trackings.filter(message => error_states.includes(message.status));
```

File: mail_tracking/static/src/core/chatter/chatter.esm.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        this.state = useState({
+            ...this.state,
+            showFailedMessageList: true,
+            isSearchFailedOpen: false,
+        });
+    },
+    get failed_messages() {
+        return this.state.thread?.messages.filter((message) => {
+            return message.is_failed_message;
+        });
```
Comment (chaule97) [inline mail_tracking/static/src/core/chatter/chatter.esm.js:27]:
```suggestion
        return this.state.thread?.messages?.filter(message => message.is_failed_message);
```

File: mail_tracking/static/src/core/message/message.esm.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    FailedMessageReview,
+    MessageTracking,
+};
+
+patch(Message.prototype, {
+    get failed_recipients() {
+        const error_states = ["error", "rejected", "spam", "bounced", "soft-bounced"];
+        return this.message.partner_trackings.filter((message) => {
+            return error_states.includes(message.status);
+        });
```
Comment (chaule97) [inline mail_tracking/static/src/core/message/message.esm.js:19]:
```suggestion
        return this.message.partner_trackings.filter(message => error_states.includes(message.status));
```

File: mail_tracking/static/src/components/failed_message/failed_message.esm.js (hunk 3 above)
Comment (trisdoan) [reply mail_tracking/static/src/components/failed_message/failed_message.esm.js:97]: Thanks for the review

As the migration changes gets quite a lot and complex (at least to me), I rather keep the diff as small as possible

If you think your change is necessary, we should open a separate PR targeting 17.0, dont you think ?

Comment (chaule97) [reply mail_tracking/static/src/components/failed_message/failed_message.esm.js:97]: I think it is not necessary, I think we don't need to change

**Review by nilshamerlinck — APPROVED** [review]

**Pull request closed without merging**